plugin/plugin_health_check.go

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package plugin
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/mattermost/mattermost-server/mlog"
+)
+
+const (
+	PLUGIN_HEALTH_CHECK_INTERVAL  = 5 // minutes
+	HEALTH_CHECK_PING_FAIL_LIMIT  = 3
+	HEALTH_CHECK_RESTART_LIMIT    = 3
+	HEALTH_CHECK_DISABLE_DURATION = 60 // minutes
+)
+
+type PluginHealthCheckJob struct {
+	cancel    chan struct{}
+	cancelled chan struct{}
+	env       *Environment
+}
+
+type PluginHealthStatus struct {
+	Crashed        bool
+	failTimeStamps []time.Time
+	lastError      error
+}
+
+// InitPluginHealthCheckJob starts a new job if one is not running and is set to enabled, or kills an existing one if set to disabled.
+func (env *Environment) InitPluginHealthCheckJob(enableJob *bool) {
+	var enable bool
+	if enableJob != nil {
+		enable = *enableJob
+	} else {
+		enable = true
+	}
+
+	// Config is set to enable. No job exists, start a new job.
+	if enable && env.pluginHealthCheckJob == nil {
+		job := newPluginHealthCheckJob(env)
+		env.pluginHealthCheckJob = job
+		job.Start()
+	}
+
+	// Config is set to disable. Job exists, kill existing job.
+	if !enable && env.pluginHealthCheckJob != nil {
+		env.pluginHealthCheckJob.Cancel()
+		env.pluginHealthCheckJob = nil
+	}
+}
+
+// Start continuously runs health checks on all active plugins, on a timer.
+func (job *PluginHealthCheckJob) Start() {
+	interval := time.Duration(PLUGIN_HEALTH_CHECK_INTERVAL) * time.Minute
+	mlog.Debug(fmt.Sprintf("Plugin health check job starting. Sending health check pings every %v minutes.", interval))
+
+	go func() {
+		defer close(job.cancelled)
+
+		env := job.env
+		activePlugins := env.Active()
+
+		for _, plugin := range activePlugins {
```

Comment:
Maybe it would be more clear that this is one time code if it was outside the `go func()`.